include/unifex/await_transform.hpp

```diff
@@ -0,0 +1,186 @@
+      op_->state_ = state::value;
+      continuation_.resume();
+    }
+
+    void set_error(std::exception_ptr eptr) && noexcept {
+      unifex::activate_union_member(op_->exception_, std::move(eptr));
+      op_->state_ = state::exception;
+      continuation_.resume();
+    }
+    
+    void set_done() && noexcept {
+      op_->state_ = state::done;
+      continuation_.resume();
+    }
+
+    template(typename CPO)
+      (requires is_receiver_query_cpo_v<CPO> AND is_callable_v<CPO, const Promise&>)
+    friend auto tag_invoke(CPO cpo, const receiver& r)
+        noexcept(is_nothrow_callable_v<CPO, const Promise&>)
+        -> callable_result_t<CPO, const Promise&> {
+      const Promise& p = r.continuation_.promise();
+      return std::move(cpo)(p);
+    }
+
+    template <typename Func>
+    friend void
+    tag_invoke(tag_t<visit_continuations>, const receiver& r, Func&& func) {
+      visit_continuations(r.continuation_.promise(), (Func&&)func);
+    }
+
+  private:
+    _as_awaitable* op_;
+    coro::coroutine_handle<Promise> continuation_;
+  };
+
+public:
+  explicit _as_awaitable(Sender&& sender, coro::coroutine_handle<Promise> h)
+    noexcept(is_nothrow_connectable_v<Sender, receiver>)
+  : op_(unifex::connect((Sender&&)sender, receiver{this, h}))
+  {}
+
+  ~_as_awaitable() {
+    switch(state_) {
+    case state::value:
+      unifex::deactivate_union_member(value_);
+      break;
+    case state::exception:
+      unifex::deactivate_union_member(exception_);
+      break;
+    default:;
+    }
+  }
+
+  bool await_ready() const noexcept {
+    return false;
+  }
+
+  void await_suspend([[maybe_unused]] coro::coroutine_handle<Promise> h) noexcept {
+    unifex::start(op_);
+  }
+
+  std::optional<value_t> await_resume() {
+    switch (state_) {
+    case state::value:
+      return std::move(value_).get();
+    case state::done:
+      return std::nullopt;
+    default:
+      assert(state_ == state::exception);
+      std::rethrow_exception(exception_.get());
+    }
+  }
+
+private:
+  state state_ = state::empty;
+  union {
+    manual_lifetime<value_t> value_;
+    manual_lifetime<std::exception_ptr> exception_;
+  };
+  connect_result_t<Sender, receiver> op_;
+};
+
+struct _fn {
+  // Call custom implementation if present.
+  template(typename Promise, typename Value)
+    (requires tag_invocable<_fn, Promise&, Value>)
+  auto operator()(Promise& promise, Value&& value) const
+    noexcept(is_nothrow_tag_invocable_v<_fn, Promise&, Value>)
+    -> tag_invoke_result_t<_fn, Promise&, Value> {
+    return unifex::tag_invoke(_fn{}, promise, (Value&&)value);
+  }
+
+  // Default implementation.
+  template(typename Promise, typename Value)
+    (requires (!tag_invocable<_fn, Promise&, Value>))
+  decltype(auto) operator()(Promise& promise, Value&& value) const {
+    if constexpr (detail::_awaitable<Value>) {
+      return (Value&&)value;
+    } else if constexpr (unifex::sender<Value>) {
```

Comment:
It's irking me that there are two `return (Value&&)value;` lines.  What do you think of collapsing this `if`?
```suggestion
    if constexpr (unifex::sender<Value> && !detail::_awaitable<Value>) {
```

Reply:
It's not the same, actually. When expressed as two separate `if constexpr` branches, if `detail::_awaitable<Value>` is true, the compiler never instantiates `unifex::sender<Value>`. This is important.

Reply:
What do you think about dropping some comments in the code to explain that?  It seems like you folks are breaking ground on how to build template libraries while breaking ground on asynchrony; it might be valuable to document what you're learning.

Reply:
A comment to that effect would be a very nice addition. And we should probably write up somewhere how we use `tag_invoke` and extreme namespace isolation as a poor-man's "[deducing `this`](http://www.open-std.org/jtc1/sc22/wg21/docs/papers/2020/p0847r4.html) to dramatically speed up compile times. There's just so much to do and so little time. We're severely understaffed.

Reply:
Comment has been added.